dali/operators/reader/loader/filesystem.cc

```diff
@@ -47,168 +47,116 @@ std::string join_path(const std::string &dir, const std::string &path) {
     return dir + dir_sep + path;
 }
 
-inline void assemble_file_list(std::vector<std::pair<std::string, int>> &file_label_pairs,
-                               const std::string &path, const std::string &curr_entry, int label,
-                               const std::vector<std::string> &filters,
-                               const bool case_sensitive_filter) {
-  std::string curr_dir_path = join_path(path, curr_entry);
-  DIR *dir = opendir(curr_dir_path.c_str());
-
-  dirent *entry;
-
-  while ((entry = readdir(dir))) {
-#ifdef _DIRENT_HAVE_D_TYPE
-    /*
-     * we support only regular files and symlinks, if FS returns DT_UNKNOWN
-     * it doesn't mean anything and let us validate filename itself
-     */
-    if (entry->d_type != DT_REG && entry->d_type != DT_LNK &&
-        entry->d_type != DT_UNKNOWN) {
-      continue;
-    }
-#endif
-    std::string rel_path = join_path(curr_entry, std::string{entry->d_name});
-    for (auto &filter : filters) {
-      if (fnmatch(filter.c_str(), entry->d_name, case_sensitive_filter ? 0 : FNM_CASEFOLD) == 0) {
-        file_label_pairs.emplace_back(rel_path, label);
-        break;
-      }
-    }
-  }
-  closedir(dir);
-}
-
-vector<std::pair<string, int>> traverse_directories(const std::string &file_root,
-                                                    const std::vector<std::string> &filters,
-                                                    const bool case_sensitive_filter) {
+bool list_subdirectories(std::vector<std::string>& subdirs, const std::string &parent_dir) {
   // open the root
-  DIR *dir = opendir(file_root.c_str());
-
-  DALI_ENFORCE(dir != nullptr,
-      "Directory " + file_root + " could not be opened.");
+  DIR *dir = opendir(parent_dir.c_str());
+  if (dir == nullptr)
+    return false;
```

Comment:
Looking at the usage, wouldn't it be better to throw here instead of returning a simple boolean success value? The error being reported is always the same (code duplication) and we lose information about the source of failure.